File: leetcode/solutions/python/26_removeDuplicatesFromSortedArray.py
# Source: https://leetcode.com/problems/remove-duplicates-from-sorted-array/
# Problem #: 26
# Author: Lily Yang
# Date: May 30, 2020
# Notes: https://github.com/lilyyanglt/leetcode/blob/master/notes/26.md


class Solution(object):
    def removeDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        try:
            nums_copy = list(nums)
        except TypeError as e:
            print(e)
            return 0
        else:
            last_element_index = len(nums) - 1

            for index, value in enumerate(nums_copy):
                if index != last_element_index and value == nums_copy[index + 1]:
                    nums.remove(value)

            return len(nums)


        # ALTERNATE SOLUTION
        # unique_count = 1
        # last_index = len(nums) - 1
        # for index, value in enumerate(nums):
        #     if index != last_index and value != nums[index + 1]:
        #         unique_count += 1
        #
        # return unique_count


def main():
    solution = Solution()

    print(solution.removeDuplicates([1, 1, 1, 1, 1, 2, 2, 2, 3]))


if __name__ == "__main__":
    main()

File: leetcode/solutions/cpp/26_removeDuplicatesFromSortedArray.cpp
/* Link: https://leetcode.com/problems/remove-duplicates-from-sorted-array/
 * Problem #: 26
 * Author: lily yang
 * Date: Aug 27, 2020

 * Approach:
  - I am using a position pointer to always point to the position I need to update if the values at nums[i] != nums[i+1].
  - If the values at nums[i] != nums[i+1]
    - I need to check if pos == i + 1, if yes, it means that there hasn't been any duplicates, so I don't do anything here other than moving incrementing the pos index
    - If pos is less than i + 1, then I know for sure there's been duplicates in the front of the array, so update nums[pos] = nums[i+1]
  - ultimately, my pos will be the total unique elements in the array

*/ 

class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return 0;
        if (n == 1) return 1;
        
        // starting pos for the update is 1 because the first value will always be in the right spot
        // since it's sorted
        int pos = 1;
        for (int i = 0; i < n-1; i++) {
            if (nums[i] != nums[i+1]) {
                if (pos < i + 1) {
                    nums[pos] = nums[i+1];
                }
                pos++;
            }
        }
        
        return pos;
    }
};